Write blog post about faster `rand` generators

RaimoNiskanen opened the pull request.

RaimoNiskanen committed:
Write blog post about faster `rand` generators

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,819 @@
+any size (bignum) range, non-overlapping sequences,
+generating floats, normal distribution floats, etc.
+Many of those features are implemented through
+a plug-in framework, whith a performance cost.
```
```suggestion
a plug-in framework, with a performance cost.
```

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,819 @@
+
+The different algorithms offered by the `rand` module are selected
+to have excellent statistical quality and to perform well
+in serious PRNG tests, (see section [PRNG tests]).
```
```suggestion
in serious PRNG tests (see section [PRNG tests]).
```

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,819 @@
+* [Write a BIF]
+* [Write a NIF]
+* [Use the system time]
+* [Hash an "unique" value]
```
```suggestion
* [Hash a "unique" value]
```
(The "u" is **pronounced** with a consonant sound; thus "a" instead of "an".)

RaimoNiskanen replied:
Although I know the rule, when I get tired i sometimes mistake a fricative for a vowel...

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,819 @@
+
+The overhead of code that has to implement this much of the features
+that the `rand` module already offers will easily approach
+it's 26 ns overhead, so often there is no point in
```
```suggestion
its 26 ns overhead, so often there is no point in
```

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,819 @@
+
+Generating a number as fast as possible (`rand:lcg35/1`) can be done
+in just above 3 ns, but that algorithm has got pretty bad
+statistical quality by to today's standards.  See section [PRNG tests].
```
```suggestion
statistical quality by today's standards.  See section [PRNG tests].
```

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,819 @@
+
+The suggested BIF `erlang:random_integer(Range)` offered
+no repeatability, generator state per scheduler, guaranteed
+sequence separatiton between schedulers, and high generator
```
```suggestion
sequence separation between schedulers, and high generator
```

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,819 @@
+in advance the range for the whole cache.
+
+A NIF could like a BIF also achieve good performance on a 32-bit system,
+with the same open question - platform independent numbers or performance?
```
```suggestion
with the same open question &mdash; platform independent numbers or performance?
```

RaimoNiskanen replied:
What is an mdash?

bjorng replied:
It is a horizontal line &mdash; roughly as long as an "m" used to mark a break in a sentence. See [Dash](https://en.wikipedia.org/wiki/Dash) and also compare the Swedish [Tankstreck](https://sv.wikipedia.org/wiki/Tankstreck), which is shorter but used in the same way as an  em dash in English.

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,819 @@
+
+
+
+### Hash an "unique" value
```
```suggestion
### Hash a "unique" value
```

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,819 @@
+### Write a simple PRNG
+
+To be fast, the implementation of a PRNG algorithm cannot
+execute many operations operations.  The operations have to be
```
```suggestion
execute many operations.  The operations have to be
```

bjorng commented on _posts/2022-04-20-faster-rand.md:
````diff
@@ -0,0 +1,819 @@
+    X1 = (A * X0 + C) rem P
+```
+For this to produce a sequence of numbers that appear random,
+there are a number of requirements on the constants A, C and P.
````
```suggestion
there are a number of requirements on the constants `A`, `C` and `P`.
```

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,819 @@
+not dig deeper into this field and instead use established research.
+
+There is a classical (1999) paper by Pierre L'Ecuyer:
+"Tables of linear congruential generators of different sizes
```
Is there an online version that you can link to?

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,819 @@
+than for a BIF.  Although the NIF used the same trick as the BIF to store
+the state in thread specific data it ended up with the same
+performance as `erlang:phash2(erlang:unique_integer(), Range)`,
+which is about 2..3 times slower than the BIF.
```
`..` is not an operator in Erlang or C, so it might not be familiar to all readers.

```suggestion
which is about 2 to 3 times slower than the BIF.
```

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,819 @@
+which I named `mcg35`
+
+`lcg35` is a power of 2 generator with an odd addition constant
+and a state interval of `0 =< X < 2^35`.
```
`^` is not defined in Erlang and it means XOR in C.

I suggest using proper mathematical notation like this: 0 &leq; X < 2 <sup>35</sup>.

```suggestion
and a state interval of 0 &leq; X < 2 <sup>35</sup>.
```

RaimoNiskanen replied:
The Markdown source becomes unreadable which goes against the idea of Markdown...

bjorng replied:
Yes, but this is a blog post. No one will read the Markdown source.

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,819 @@
+and a state interval of `0 =< X < 2^35`.
+
+`mcg35` is a prime modulus multiplicative generator (Lehmer generator)
+with a state interval of `1 =< X < 2^35-31`.  Since the modulus,
```
```suggestion
with a state interval of 1  &leq; X < 2 <sup>35</sup>-1.  Since the modulus,
```

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,819 @@
+Speed over quality?
+-------------------
+
+The Pseudo Random Number Generators implemented in
```
```suggestion
The Pseudo Random Number Generators (PRNGs) implemented in
```

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,819 @@
+### PRNG tests
+
+There are test frameworks that tests the statistical properties
+of PRNG:s, such as the TestU01 framework, or PracRand.
```
Can you link to them?

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,819 @@
+
+The speed of the newly implemented `lcg53` and `mcg53` algorithms
+is much thanks to the recent [type-based optimizations] in the compiler
+and the Just-In-Time compliling .BEAM code loader.
```
```suggestion
and the Just-In-Time compiling BEAM code loader.
```

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,819 @@
+### With type-based optimization
+
+When the compiler can figure out type information about the arguments
+it can emit more effective code.  One would like to add a guard
```
```suggestion
it can emit more efficient code.  One would like to add a guard
```

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,819 @@
+But adding a redundant input bit mask to the Erlang code puts the compiler
+on the right track.  This is a kludge, and will only be used
+until the compiler has been improved to deduce the same information
+from a guard instead:
```
```suggestion
from a guard instead.
```

bjorng commented on _posts/2022-04-20-faster-rand.md:
````diff
@@ -0,0 +1,819 @@
+    {gc_bif,'+',{f,0},1,[{tr,{x,0},number},{integer,15366142135}],{x,0}}.
+    {gc_bif,'band',{f,0},1,[{tr,{x,0},number},{integer,34359738367}],{x,0}}.
+```
+when loaded by the JIT (x86) (erl +JDdump true) the machine code becomes:
````
```suggestion
when loaded by the JIT (x86) (`erl +JDdump true`) the machine code becomes:
```

bjorng commented on _posts/2022-04-20-faster-rand.md:
````diff
@@ -0,0 +1,819 @@
+```
+when loaded by the JIT (x86) (erl +JDdump true) the machine code becomes:
+```
+# i_test_yield
````
I suggest omitting the `i_test_yield` instruction and only show the instructions relevant to the example.

RaimoNiskanen replied:
Ok.  My mistake - I thought that was part of the multiplication...

bjorng commented on _posts/2022-04-20-faster-rand.md:
````diff
@@ -0,0 +1,819 @@
+    cmp edi, 15
+    short jne L1891
+```
+Above was a test for if X0 is a small integer.
````
```suggestion
Above was a test for `{x,0}` being a small integer.
```

bjorng commented on _posts/2022-04-20-faster-rand.md:
````diff
@@ -0,0 +1,819 @@
+    or rax, 15
+    short jmp L1890
+```
+If X0 was not a small integer the following fallback code is used
````
I suggest using `{x,0}` (the same as in the BEAM code shown earlier) instead of `X0`.

```suggestion
If `{x,0}` was not a small integer the following fallback code is used
```

bjorng commented on _posts/2022-04-20-faster-rand.md:
````diff
@@ -0,0 +1,819 @@
+            {x,0}}.
+```
+Note the `{t_integer,Range}` type information that instead of just declaring
+that the argument is a number, now states that it is an integer on
````
```suggestion
that the argument is a number, now states that it is an integer in
```

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,819 @@
+In this section `lcg35_inline`, `exsp_inline` and `system_time`
+use bit operations such as `X band 16#ffffffff` or `X bsr 3`
+to achieve the desired range, and now we see that this is
+about 5..10 ns faster than for the `rem` operation
```
```suggestion
about 5 to 10 ns faster than for the `rem` operation
```

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,819 @@
+The Pseudo Random Number Generators implemented in
+the `rand` module offers many useful features such as
+repeatable sequences, non-biased range generation,
+any size (bignum) range, non-overlapping sequences,
```
Here you mention "bignum" without defining it. I suggest that you delete this mention and defines "bignum" a little bit later.
 
```suggestion
any size range, non-overlapping sequences,
```

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,819 @@
+to have excellent statistical quality and to perform well
+in serious PRNG tests, (see section [PRNG tests]).
+
+Most of these algorithms are designed for machines with
```
Somewhere here could be a good place to define what bignums are and also explain why `rand` uses 58-bit integers (why not 59-bit integers).

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,819 @@
+that boasts sub-ns speed in C needs 17 ns in Erlang.
+
+32-bit Erlang is a sad story in this regard.  The bignum limit
+on such an Erlang system is so low that designing a PRNG
```
How low?

RaimoNiskanen replied:
You tell me ;-)  `erlang:phash2/1` returns 27 bits.  Is 2^27 - 1 the upper limit for a positive immediate?

bjorng replied:
Yes.

RaimoNiskanen replied:
Great!  Then we would have to use 26-bit integers.  An LCG would have to have a multiplication constant of 13 bits.  So size and period about 8192.  Ridiculous!

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,819 @@
+Pseudo Random Number Generators in the `rand` module are overkill.
+This blog post will dive in to new additions to the
+said module, how the Just-In-Time compiler optimizes them,
+known tricks, and and tries to compare these apples and potatoes.
```
```suggestion
known tricks, and tries to compare these apples and potatoes.
```

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,819 @@
+to handle than immediate integers.
+
+The algorithms that perform well in Erlang are the ones
+that have been redesigned to work on 58-bit words.
```
```suggestion
that have been redesigned to work on 58-bit integers.
```

bjorng reviewed (commented):
Overall looks good to me.

RaimoNiskanen commented:
Thank you for a thorough review!

RaimoNiskanen committed:
Update after feedback

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,841 @@
+an infinite period, since the time it will take for it to repeat
+is assumed to be longer than the Erlang node will survive.
+
+For the new fast generators the period it is about 2<sup>35</sup>`,
+and for the good ones in `rand` it is at least `2<sup>116</sup>&nbsp;-&nbsp;1,
+which is a huge difference and also much more than what can be consumed
```
Remove stray backticks which caused strange formatting:

```suggestion
For the new fast generators the period it is about 2<sup>35</sup>,
and for the good ones in `rand` it is at least 2<sup>116</sup>&nbsp;-&nbsp;1,
which is a huge difference and also much more than what can be consumed
```

RaimoNiskanen replied:
Bummer!  I missed some!

RaimoNiskanen committed:
Remove stray backticks

RaimoNiskanen committed:
Refer to BEAM X registers  as `{x,_}` registers

RaimoNiskanen committed:
Rewrite for `lcg35` and `mcg35` -> `mwc59`

RaimoNiskanen committed:
Change from `lcg35` and `mcg35` to `mwc59`

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,840 @@
+
+### Spectral score
+
+The specral score of a generator, measures how a much a sequence of numbers
```
```suggestion
The spectral score of a generator, measures how a much a sequence of numbers
```

RaimoNiskanen replied:
Also: "how a much a" -> "how much a"

bjorng commented on _posts/2022-04-20-faster-rand.md:
````diff
@@ -0,0 +1,840 @@
+    short jne L2271
+```
+Above was a test if `{x,0}` is a small integer and if not
+the fallback at `L1816` is called to handle any term.
````
```suggestion
the fallback at `L2271` is called to handle any term.
```

bjorng commented on _posts/2022-04-20-faster-rand.md:
````diff
@@ -0,0 +1,840 @@
+```
+Above was a test if `{x,0}` is a small integer and if not
+the fallback at `L1816` is called to handle any term.
+Here follows the machine code shift left, Erlang `bsl 32`,
````
```suggestion
Here follows the machine code for shift right, Erlang `bar 32`,
```

bjorng commented on _posts/2022-04-20-faster-rand.md:
````diff
@@ -0,0 +1,840 @@
+It is a well researched PRNG, and can be implemented
+like this (in Erlang):
+``` erlang
+    C   = CX0 bsl B,
````
```suggestion
    C   = CX0 bsr B,
```

bjorng commented on _posts/2022-04-20-faster-rand.md:
```diff
@@ -0,0 +1,840 @@
+Above was a test if `{x,0}` is a small integer and if not
+the fallback at `L1816` is called to handle any term.
+Here follows the machine code shift left, Erlang `bsl 32`,
+x86 `sar rax 32`, and a skip over the fallback code.
```
```suggestion
x86 `sar rax, 32`, and a skip over the fallback code.
```

bjorng reviewed (commented):
Found a few typos, otherwise looks good.

RaimoNiskanen committed:
Change `mwc59` multiplier

RaimoNiskanen committed:
Fix typos from revies

RaimoNiskanen committed:
Fix after spell checker

RaimoNiskanen committed:
Polish details

RaimoNiskanen committed:
Update for new generator constants

RaimoNiskanen committed:
Update to follow source code development

RaimoNiskanen committed:
Write about the MWC development

Write a new section treating how and why the `mwc59` generator
got its parameters and scramblers.

Update the mesaurement results after code freeze.

RaimoNiskanen committed:
Work on the links

Create a table of content.
Add more external links.

RaimoNiskanen committed:
Update after feedback

RaimoNiskanen committed:
Update after new measurements

Refresh the measurement results after merge to 'master'.
Add non-breaking spaces in some expressions.

RaimoNiskanen committed:
Review after read-through

Move duplicate text from [Write a simple PRNG] to [Implementing a PRNG].
Polish the text in general.

RaimoNiskanen committed:
Rename blog post date
